### ISSUE TITLE
Logging into MeshCentral via PHP using a URL string

### ISSUE THREAD

**Issue by TheMonk76** [Ylianst/MeshCentral#1257]
Dear Ylianst,

First of all, let me thank you for the awesome application you call "MEshCentral". It made our life easier. I wander if i can login to MeshCentral via tweaking the URL. That is to say, can i for example login like this --> http://serverFQDN?username=user1&password=usrpass

P.S. assuming that there is a user called user1 and his password is usrpass.

Would like to know how can i do such a task if it can be done?

FYI.. i am using PHP to create a splash screen that acts like a "single sign on" place for our companies users.

Much obliged,

Mouhab

**Comment by ryanblenis** [comment]
Hi @TheMonk76,

Putting the username and password in plaintext would be a security issue. 

If you're a PHP developer, take a look at the createLoginToken.ashx function in webserver.js and you should be able to create a PHP function to mimic the same functionality. That would allow you to create an encrypted authentication token that can be used to log in via URL for a given username/password.

I'm not sure if MC's built-in LDAP functionality is SSO (Single Sign On) or only SSI (Single Sign In), but Ylian may comment if that may be a better approach.

**Comment by TheMonk76** [comment]
@ryanblenis Thank you for the advice about the security concern. I am new to coding and still learning. I was given that task by my senior and wanted to know if there is a way to communicate with the MeshCentral and ask the server to log someone. Like i have server X and it hosts my PHP code that has a splash screen with all the links to our company cloud apps (e.g. MeshCentral). My senior wanted me to add a link that can log someone based on his saved MeshCentral username & password. So i guess what i need is SSI, right?

Thanks again for the guidance here and i will try to see how i can utilize createLoginToken.ashx function in webserver.js

**Comment by ryanblenis** [comment]
SSI is Single Sign In, where you log in once (presumably your computer, or a website) and then you're "logged in" to all your web services / accounts. Single Sign On would be that you have the same credentials to everything, so you can go to O365 / your Windows workstation / MeshCentral and use the same username and password (but still be prompted for them) at each service.

On a second look, you don't even need to know the users password (and rightfully so) to create them a login cookie. Just mirror the cookie creation to create a JSON object like:

`{ a: 3, u: 'user//adminusernamehere', time: 1588101882000, dtime: 659145 }`

`a: 3` is what allows the cookie to be used for login
`u: username` is the user to log in as
`time` is the unix epoch time of the cookie creation (good for 4 hours)

If you can just create an AESGCM encoded cookie using the same key that your MeshCentral install is using you should be able to create a link to log your users in securely by pointing them to https://your.url.com/?login=COOKIEDATA

**Comment by Ylianst** [comment]
Two options for this. First, as @ryanblenis described, you can generate a login token and use that to login. The token is time limited and you need the server's secret key to generate one. There is section 14.1 "Login Token" in the [MeshCentral User's Guide](http://info.meshcentral.com/downloads/MeshCentral2/MeshCentral2UserGuide-0.2.7.pdf).

The next option is to take a look at MeshCtrl.js. This is a command line tool that comes with MeshCentral. In your MeshCentral server, try:

```
node node_modules/meshcentral/meshctrl
```

This tool allows you to create accounts and do lots of other operations. You can also copy the "meshctrl.js" on a different computer and use it to login to MeshCentral. If you look at the code, you could also do the same in Python. You need to connect to MeshCentral using a websocket and you send your username and password as a command within websocket.

Using username/password in the URL is very bad for security because URL's are often logged at many places. So, best to use one of the techniques above. Hope it helps.

**Comment by TheMonk76** [comment]
@Ylianst Many thanks for your valuable insights. I wonder what is the best practice as to how to communicate the "Login Token Key" via PHP to the MeshCentral Server in order to get a response with the "Login Token" back from the server?
Thanks for the insight.

**Comment by TheMonk76** [comment]
> SSI is Single Sign In, where you log in once (presumably your computer, or a website) and then you're "logged in" to all your web services / accounts. Single Sign On would be that you have the same credentials to everything, so you can go to O365 / your Windows workstation / MeshCentral and use the same username and password (but still be prompted for them) at each service.
> 
> On a second look, you don't even need to know the users password (and rightfully so) to create them a login cookie. Just mirror the cookie creation to create a JSON object like:
> 
> `{ a: 3, u: 'user//adminusernamehere', time: 1588101882000, dtime: 659145 }`
> 
> `a: 3` is what allows the cookie to be used for login
> `u: username` is the user to log in as
> `time` is the unix epoch time of the cookie creation (good for 4 hours)
> 
> If you can just create an AESGCM encoded cookie using the same key that your MeshCentral install is using you should be able to create a link to log your users in securely by pointing them to https://your.url.com/?login=COOKIEDATA

Much appreciated @ryanblenis

**Comment by ryanblenis** [comment]
> I wonder what is the best practice as to how to communicate the "Login Token Key" via PHP to the MeshCentral Server in order to get a response with the "Login Token" back from the server?
> Thanks for the insight.

The createLoginToken.ashx is blocked on servers with an SSL certificate (e.g. only usable on test instances) because it is an attack vector someone could use to brute-force user/password combinations, so if you're referring to using that for the token creation, you wouldn't be able to. If you had a local, locked down instance of MC running locally without a trusted cert and "happened" to have the same key on it as the production server, you could get around this limitation. You can simply wget, curl, or fetch the response as it comes back as straight text.
Of course, this method still assumes you know the username and password to each account (which you shouldn't) to generate. Recreating the function and only knowing the username would be proper approach.

If, on the other hand you're referring to logging the user in once you've got that info, just use a location redirect to the MC url and add the variable you stored the token in.

**Comment by Ylianst** [comment]
Ideally, you would want to port the encodeCookie() method to PHP so that if you have the server secret, you can generate your own login tokens. Another option is to cut & paste the encodeCookie() method into it's own small nodejs application and call it form PHP passing in the server key and the username you want to login it as.

The entire idea of the cookie was for a separate server to be able to generate it as needed, not to request the MeshCentral generate it. Hope that helps.

**Comment by Ylianst** [comment]
Closing this open. Feel free to reopen or post a new issue if needed.
